sisl/io/orca/output.py

```diff
@@ -0,0 +1,175 @@
+        itt = iter(self)
+        next(itt) # skip ---
+        A = np.empty((natoms, 2), np.float64)
+        for ia in range(natoms):
+            line = next(itt)
+            v = line.split()
+            A[ia] = float(v[-2]), float(v[-1])
+        return A
+
+
+    @sile_fh_open()
+    def read_mulliken_atomic(self):
+        """ Reads the atom-resolved Mulliken charge and spin population analysis
+
+        Returns
+        -------
+        ndarray : atom-resolved charge and spin populations
+        """
+
+        natoms = self._natoms
+
+        f = self.step_to("MULLIKEN ATOMIC CHARGES AND SPIN POPULATIONS")[0]
+        if not f:
+            return None
+
+        return self._read_atomic_block(natoms)
+
+
+    @sile_fh_open()
+    def read_loewdin_atomic(self):
+        """ Reads the atom-resolved Loewdin charge and spin population analysis
+
+        Returns
+        -------
+        ndarray : atom-resolved charge and spin populations
+        """
+
+        natoms = self._natoms
+
+        f = self.step_to("LOEWDIN ATOMIC CHARGES AND SPIN POPULATIONS")[0]
+        if not f:
+            return None
+
+        return self._read_atomic_block(natoms)
+
+
+    def _read_orbital_block(self):
+        itt = iter(self)
+        D = PropertyDict()
+        v = next(itt).split()
+        while len(v) > 0:
+            if len(v) == 8:
+                ia = int(v[0])
+                D[(ia, v[2])] = float(v[4])
+                D[(ia, v[5])] = float(v[7])
+            elif len(v) == 6:
+                D[(ia, v[0])] = float(v[2])
+                D[(ia, v[3])] = float(v[5])
+            else:
+                D[(ia, v[0])] = float(v[2])
+            v = next(itt).split()
+        return D
+
+
+    @sile_fh_open()
+    def read_mulliken_orbitals(self, orbital=None):
+        """ Reads the orbital-resolved Mulliken charge and spin population analysis
+
+        Parameters
+        ----------
+        orbital : str, optional
+            if an orbital string is specified a ndarray is returned with the corresponding
+            values per atom, otherwise the whole orbital block is returned as dictionary
+
+        Returns
+        -------
+        ndarrays or PropertyDicts : charge and spin data
+        """
+
+        f = self.step_to("MULLIKEN REDUCED ORBITAL CHARGES AND SPIN POPULATIONS", reread=False)[0]
+        if not f:
+            return None
+
+        self.step_to("CHARGE", reread=False)
+        charge = self._read_orbital_block()
+
+        self.step_to("SPIN", reread=False)
+        spin = self._read_orbital_block()
+
+        if orbital is not None:
+            natoms = self._natoms
+            c = np.zeros(natoms, np.float64)
+            s = np.zeros(natoms, np.float64)
+            for key in charge:
+                ia, orb = key
+                if orb == orbital:
+                    c[ia] = charge[key]
+                    s[ia] = spin[key]
+            return c, s
+
+        return charge, spin
```

Comment:
OK, I implemented this with the `orbital` keyword and the user then gets a single array of shape (natoms, 2).

Without it, I think the dictionaries are simpler. The block for one atom looks like this (and similar for SPIN that comes later in the file):
```
CHARGE
  0 C s       :     3.194163  s :     3.194163
      pz      :     0.963188  p :     2.811388
      px      :     0.922070
      py      :     0.926130
      dz2     :     0.004461  d :     0.080121
      dxz     :     0.004369
      dyz     :     0.022705
      dx2y2   :     0.032393
      dxy     :     0.016192
      f0      :     0.001427  f :     0.006652
      f+1     :     0.000581
      f-1     :     0.000488
      f+2     :     0.000136
      f-2     :     0.001101
      f+3     :     0.000907
      f-3     :     0.002012
  1 C s       :     3.269684  s :     3.269684
      pz      :     0.984837  p :     2.836259
...
```
The function currently maps this into a dictionary of this form:
```
{(0, 's'): 3.194163, (0, 'pz'): 0.963188, (0, 'p'): 2.811388,
(0, 'px'): 0.92207, (0, 'py'): 0.92613, (0, 'dz2'): 0.004461,
(0, 'd'): 0.080121, (0, 'dxz'): 0.004369, (0, 'dyz'): 0.022705,
(0, 'dx2y2'): 0.032393, (0, 'dxy'): 0.016192, (0, 'f0'): 0.001427,
(0, 'f'): 0.006652, (0, 'f+1'): 0.000581, (0, 'f-1'): 0.000488,
(0, 'f+2'): 0.000136, (0, 'f-2'): 0.001101, (0, 'f+3'): 0.000907,
(0, 'f-3'): 0.002012, (1, 's'): 3.269684, (1, 'pz'): 0.984837,
(1, 'p'): 2.836259, ...
```

Reply:
A dictionary is nice for segmenting things, but this would force end-users to dissect it if wanting to plot things. Generally I don't return data-containers as dictionaries if they have shared meaning (at least I don't think I do ;)). What would be ideal is to use dataframes or something similar that lets users query shells, this is also suggested by @pfebrer. I am just worried about dataframes having lots of empty entries, whilst simpler data-instances could make it easier. Hmm...

If your read-geometry assigns orbitals to the atoms, it should be easier to figure out orbital indices etc.

Reply:
Currently read-geometry is implemented for the `property.txt` file which does not contain the orbital information...

Note also that the populations are not directly referring to the basis functions but to a projection in terms of `n,l,m` quantum numbers. For instance, in the block above was computed with a `def2-TZVP`" basis ("valence triple-zeta basis set with “new” polarization functions"), i.e., this projection block is not resolved per basis orbital.

Reply:
hmmm... ok. I see...

Reply:
I figured out that one can also request the population analysis in the full orbital basis and not only the reduced one (which is the default), see the keyword `reduced=True` how this can now be accessed too.